### PR TITLE
Generate test application automatically

### PR DESCRIPTION
**Pull request by maximehuran**
## Use Makefile to generate test application

- Add `application` target wich will initialize a test application from `sylius-standard`
- Add `dist` folder with some file to copy in our test application
- Ignore totally `tests/Application` folder in git
- Ignore totally `tests/Application` folder in PHPCS
- Use custom folder for PHPUnit tests, in order to avoid all unit tests in `tests/Application/vendor` !

### CODE REVIEW

File: .github/workflows/recipe.yaml [hunk 1]
```diff
@@ -46,25 +44,24 @@ jobs:
           docker run --rm --name flex -d -v $PWD/plugin/recipes:/var/www/flex/var/repo/private/monsieurbiz/sylius-cms-page-plugin -p 80:80 docker.pkg.github.com/monsieurbiz/docker/symfony-flex-server:latest contrib official
           docker ps
 
+      - run: mkdir -p /home/runner/{.composer/cache,.config/composer}
+
```
Comment (jacquesbh) [inline .github/workflows/recipe.yaml:48]: @maximehuran Isn't it blocking the cache if this `mkdir` is done before next step?

Isn't it better to keep it after?

File: Makefile [hunk 2]
```diff
@@ -1,72 +1,111 @@
...
-reset: docker.down ## Stop docker and remove dependencies
+reset: ## Stop docker and remove dependencies
+	${MAKE} docker.down || true
 	rm -rf ${APP_DIR}/node_modules ${APP_DIR}/yarn.lock
+	rm -rf ${APP_DIR}
 	rm -rf vendor composer.lock
 .PHONY: reset
 
-dependencies: vendor node_modules ## Setup the dependencies
+dependencies: composer.lock node_modules ## Setup the dependencies
```
Comment (jacquesbh) [inline Makefile:31]: If we have a composer.lock but no vendor/ directory then nothing will happen…

File: Makefile [hunk 3]
```diff
@@ -1,72 +1,111 @@
...
 .PHONY: dependencies
 
 .php-version: .php-version.dist
-	cp .php-version.dist .php-version
-	(cd ${APP_DIR} && ln -sf ../../.php-version)
+	rm -f .php-version
+	ln -s .php-version.dist .php-version
```
Comment (jacquesbh) [inline Makefile:36]: Please copy, do not use a symlink.

File: Makefile [hunk 4]
```diff
@@ -1,72 +1,111 @@
...
 
-vendor: composer.lock ## Install the PHP dependencies using composer
-ifdef GITHUB_ACTIONS
-	${COMPOSER} install --prefer-dist
-else
-	${COMPOSER} install --prefer-source
-endif
+php.ini: php.ini.dist
+	rm -f php.ini
+	ln -s php.ini.dist php.ini
```
Comment (jacquesbh) [inline Makefile:40]: Same here, could you use a copy and not a symlink please?>

File: Makefile [hunk 5]
```diff
@@ -1,72 +1,111 @@
...
+
+setup_application:
+	rm -f ${APP_DIR}/yarn.lock
+	(cd ${APP_DIR} && ${COMPOSER} config repositories.plugin '{"type": "path", "url": "../../"}')
+	(cd ${APP_DIR} && ${COMPOSER} config extra.symfony.allow-contrib true)
+	(cd ${APP_DIR} && ${COMPOSER} config minimum-stability dev)
+	(cd ${APP_DIR} && ${COMPOSER} require --no-scripts --no-progress --no-install --no-update monsieurbiz/${PLUGIN_NAME}="*@dev")
+	$(MAKE) apply_dist
+	$(MAKE) ${APP_DIR}/.php-version
+	$(MAKE) ${APP_DIR}/php.ini
```
Comment (jacquesbh) [inline Makefile:75]:
```suggestion
	$(MAKE) apply_dist ${APP_DIR}/.php-version ${APP_DIR}/php.ini
```

File: Makefile [hunk 6]
```diff
@@ -1,72 +1,111 @@
...
+	(cd ${APP_DIR} && ${COMPOSER} install)
+
+${APP_DIR}/docker-compose.yaml:
+	rm -f ${APP_DIR}/docker-compose.yml
+	rm -f ${APP_DIR}/docker-compose.yaml
+	ln -s ../../docker-compose.yaml.dist ${APP_DIR}/docker-compose.yaml
```
Comment (jacquesbh) [inline Makefile:81]: Same, a copy is fine.

File: Makefile [hunk 7]
```diff
@@ -1,72 +1,111 @@
...
+.PHONY: ${APP_DIR}/docker-compose.yaml
+
+${APP_DIR}/.php-version: .php-version
+	(cd ${APP_DIR} && ln -sf ../../.php-version)
+
+${APP_DIR}/php.ini: php.ini
+	(cd ${APP_DIR} && ln -sf ../../php.ini)
```
Comment (jacquesbh) [inline Makefile:88]: Copy… copy…

File: Makefile [hunk 8]
```diff
@@ -1,72 +1,111 @@
...
+
+apply_dist:
+	ROOT_DIR=$(shell dirname $(realpath $(firstword $(MAKEFILE_LIST)))); \
+	for i in `cd dist && find . -type f`; do \
+		FILE_PATH=`echo $$i | sed 's|./||'`; \
+		FOLDER_PATH=`dirname $$FILE_PATH`; \
+		echo $$FILE_PATH; \
+		(cd ${APP_DIR} && rm -f $$FILE_PATH); \
+		(cd ${APP_DIR} && mkdir -p $$FOLDER_PATH); \
+		(cd ${APP_DIR} && ln -s $$ROOT_DIR/dist/$$FILE_PATH $$FILE_PATH); \
```
Comment (jacquesbh) [inline Makefile:98]: Symlinks are fine here.

File: composer.json [hunk 9]
```diff
@@ -31,47 +31,46 @@
...
-        "psr-4": {
-            "App\\": "tests/Application/src/",
-            "Tests\\MonsieurBiz\\SyliusCmsPagePlugin\\": "tests/"
-        }
-    },
-    "extra": {
-        "branch-alias": {
-            "dev-master": "1.0-dev"
-        },
-        "config-dir": "./tests/Application/config/"
```
Comment (jacquesbh) [inline composer.json:66]: Isn't this useful?

File: composer.json [hunk 10]
```diff
@@ -31,47 +31,46 @@
...
         "phpstan": "phpstan analyse -c phpstan.neon src/",
-        "phpmd": "phpmd --exclude Migrations/* src/ ansi phpmd.xml"
+        "phpmd": "phpmd --exclude Migrations/* src/ ansi phpmd.xml",
+        "phpunit": "phpunit",
+        "phpspec": "phpspec run"
+    },
+    "extra": {
+        "symfony": {
+            "require": "^4.4"
+        },
```
Comment (jacquesbh) [inline composer.json:71]: Is that required?

File: phpstan.neon [hunk 11]
```diff
@@ -7,7 +7,6 @@ parameters:
     paths:
         - %rootDir%/src/
 
-    reportUnmatchedIgnoredErrors: false
     checkMissingIterableValueType: false
```
Comment (jacquesbh) [inline phpstan.neon:10]:
```suggestion
    checkMissingIterableValueType: false
    checkGenericClassInNonGenericObjectType: false
```

File: src/Context/LastChanceLocaleContext.php [hunk 12]
```diff
@@ -33,7 +34,7 @@ public function __construct(
 
     public function getLocaleCode(): string
     {
-        $request = $this->requestStack->getMainRequest();
+        $request = $this->requestStack->getMasterRequest();
```
Comment (jacquesbh) [inline src/Context/LastChanceLocaleContext.php:37]: Hum… really?

File: src/Routing/RequestContext.php [hunk 13]
```diff
@@ -71,25 +63,15 @@ public function __construct(
...
 
-    /**
-     * @param Request $request
-     *
-     * @return bool
-     */
     public function checkPageSlug(Request $request): bool
     {
-        if ($request !== $this->requestStack->getMainRequest()) {
+        if ($request !== $this->requestStack->getMasterRequest()) {
```
Comment (jacquesbh) [inline src/Routing/RequestContext.php:68]: Really?

File: symfony.lock [hunk 14]
```diff
@@ -1,1150 +0,0 @@
-{
```
Comment (jacquesbh) [inline symfony.lock:1]: Did you remove this file? Don't we want to keep it?

**Review by jacquesbh — CHANGES_REQUESTED** [review]

**Review by delyriand — APPROVED** [review]

**Pull request merged**